File: src/adapters/mock.ts
import { SahajError } from '../errors';
import { buildDepositFixtures } from '../sandbox/fixtures';
import { type SandboxScenario, resolveScenario } from '../sandbox/vua';
import type {
  AAAdapter,
  AdapterConsent,
  AdapterSession,
  ConsentRequest,
  ConsentStatus,
  FetchedFip,
  SandboxControls,
} from './adapter';

/**
 * The sandbox adapter. It runs the whole consent-to-data loop in memory with no
 * credentials and no network, driving each magic-VUA scenario through the same
 * state machine a real AA would. This is the wedge: the loop works in minutes,
 * offline, before anyone holds an FIU licence.
 */

export interface MockAdapterOptions {
  /** Approve (or reject/expire, per the VUA) as soon as the consent is created. Default true. */
  readonly autoApprove?: boolean;
  /** Seed for the deterministic fixtures. Default 1. */
  readonly seed?: number;
}

interface MockConsentState {
  readonly id: string;
  readonly mobile: string;
  readonly scenario: SandboxScenario;
  readonly fiTypes: ConsentRequest['fiTypes'];
  readonly createdAt: string;
  readonly expiresAt: string;
  status: ConsentStatus;
}

const SANDBOX_CLOCK_MILLIS = Date.UTC(2026, 5, 1, 9, 0, 0); // fixed sandbox "now" for reproducible timestamps

/** What a given scenario resolves to once the human step happens. */
function approvalOutcome(scenario: SandboxScenario): ConsentStatus {
  if (scenario === 'reject') {
    return 'REJECTED';
  }
  if (scenario === 'expire') {
    return 'EXPIRED';
  }
  return 'ACTIVE';
}

export class MockAdapter implements AAAdapter, SandboxControls {
  readonly name = 'sandbox';

  private readonly autoApprove: boolean;
  private readonly seed: number;
  private readonly consents = new Map<string, MockConsentState>();
  private readonly consentIdByIdempotencyKey = new Map<string, string>();
  private consentCounter = 0;

  constructor(options: MockAdapterOptions = {}) {
    this.autoApprove = options.autoApprove ?? true;
    this.seed = options.seed ?? 1;
  }

  async createConsent(request: ConsentRequest): Promise<AdapterConsent> {
    if (request.idempotencyKey) {
      const existingId = this.consentIdByIdempotencyKey.get(request.idempotencyKey);
      if (existingId) {
        return this.toAdapterConsent(this.requireConsent(existingId));
      }
    }

    const scenario = resolveScenario(request.mobile);
    const id = `csnt_sandbox_${++this.consentCounter}`;
    const state: MockConsentState = {
      id,
      mobile: request.mobile,
      scenario,
      fiTypes: request.fiTypes,
      createdAt: new Date(SANDBOX_CLOCK_MILLIS).toISOString(),
      expiresAt: new Date(SANDBOX_CLOCK_MILLIS + 15 * 60 * 1000).toISOString(),
      status: this.autoApprove ? approvalOutcome(scenario) : 'PENDING',
    };
    this.consents.set(id, state);
    if (request.idempotencyKey) {
      this.consentIdByIdempotencyKey.set(request.idempotencyKey, id);
    }
    return this.toAdapterConsent(state);
  }

  async getConsentStatus(consentId: string): Promise<ConsentStatus> {
    return this.requireConsent(consentId).status;
  }

  async createSession(consentId: string): Promise<AdapterSession> {
    const state = this.requireConsent(consentId);
    if (state.scenario === 'fip_down') {
      throw new SahajError('FIP_UNAVAILABLE', { fipId: 'SBIN' });
    }
    return { id: `sess_sandbox_${state.id}`, status: 'COMPLETED' };
  }

  async fetchData(consentId: string, _sessionId: string): Promise<FetchedFip[]> {
    const state = this.requireConsent(consentId);
    if (state.scenario === 'no_accounts') {
      throw new SahajError('NO_ACCOUNTS_FOUND');
    }
    return buildDepositFixtures(state.mobile, state.scenario, this.seed).map((fip) => ({
      fipId: fip.fipId,
      fipName: fip.fipName,
      status: fip.status,
      fiType: 'DEPOSIT' as const,
      payload: fip.payload,
    }));
  }

  approve(consentId: string): void {
    const state = this.requireConsent(consentId);
    state.status = approvalOutcome(state.scenario);
  }

  reject(consentId: string): void {
    this.requireConsent(consentId).status = 'REJECTED';
  }

  private requireConsent(consentId: string): MockConsentState {
    const state = this.consents.get(consentId);
    if (!state) {
      throw new SahajError('INVALID_INPUT', { field: 'consentId' });
    }
    return state;
  }

  private toAdapterConsent(state: MockConsentState): AdapterConsent {
    return {
      id: state.id,
      status: state.status,
      redirectUrl: `https://sandbox.sahaj.local/consents/${state.id}/approve`,
      createdAt: state.createdAt,
      expiresAt: state.expiresAt,
    };
  }
}

File: python/src/sahaj/adapters/mock.py
"""The sandbox adapter. It runs the whole consent-to-data loop in memory with no
credentials and no network, driving each magic-VUA scenario through the same state
machine a real AA would. This is the wedge: the loop works in minutes, offline,
before anyone holds an FIU licence.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
from typing import Optional

from ..errors import SahajError
from ..sandbox.fixtures import build_deposit_fixtures
from ..sandbox.vua import resolve_scenario
from .base import AdapterConsent, AdapterSession, ConsentRequest, FetchedFip

_SANDBOX_CLOCK = datetime(2026, 6, 1, 9, 0, 0, tzinfo=timezone.utc)


def _iso(moment: datetime) -> str:
    return moment.isoformat().replace("+00:00", "Z")


def _approval_outcome(scenario: str) -> str:
    if scenario == "reject":
        return "REJECTED"
    if scenario == "expire":
        return "EXPIRED"
    return "ACTIVE"


@dataclass
class _MockConsentState:
    id: str
    mobile: str
    scenario: str
    fi_types: list[str]
    created_at: str
    expires_at: str
    status: str


class MockAdapter:
    name = "sandbox"

    def __init__(self, auto_approve: bool = True, seed: int = 1) -> None:
        self._auto_approve = auto_approve
        self._seed = seed
        self._consents: dict[str, _MockConsentState] = {}
        self._consent_id_by_idempotency_key: dict[str, str] = {}
        self._consent_counter = 0

    def create_consent(self, request: ConsentRequest) -> AdapterConsent:
        if request.idempotency_key:
            existing_id = self._consent_id_by_idempotency_key.get(request.idempotency_key)
            if existing_id:
                return self._to_adapter_consent(self._require_consent(existing_id))

        scenario = resolve_scenario(request.mobile)
        self._consent_counter += 1
        consent_id = f"csnt_sandbox_{self._consent_counter}"
        state = _MockConsentState(
            id=consent_id,
            mobile=request.mobile,
            scenario=scenario,
            fi_types=request.fi_types,
            created_at=_iso(_SANDBOX_CLOCK),
            expires_at=_iso(_SANDBOX_CLOCK + timedelta(minutes=15)),
            status=_approval_outcome(scenario) if self._auto_approve else "PENDING",
        )
        self._consents[consent_id] = state
        if request.idempotency_key:
            self._consent_id_by_idempotency_key[request.idempotency_key] = consent_id
        return self._to_adapter_consent(state)

    def get_consent_status(self, consent_id: str) -> str:
        return self._require_consent(consent_id).status

    def create_session(self, consent_id: str) -> AdapterSession:
        state = self._require_consent(consent_id)
        if state.scenario == "fip_down":
            raise SahajError("FIP_UNAVAILABLE", fip_id="SBIN")
        return AdapterSession(id=f"sess_sandbox_{state.id}", status="COMPLETED")

    def fetch_data(self, consent_id: str, _session_id: str) -> list[FetchedFip]:
        state = self._require_consent(consent_id)
        if state.scenario == "no_accounts":
            raise SahajError("NO_ACCOUNTS_FOUND")
        return [
            FetchedFip(
                fip_id=fip.fip_id,
                fip_name=fip.fip_name,
                status=fip.status,
                fi_type="DEPOSIT",
                payload=fip.payload,
            )
            for fip in build_deposit_fixtures(state.mobile, state.scenario, self._seed)
        ]

    def approve(self, consent_id: str) -> None:
        state = self._require_consent(consent_id)
        state.status = _approval_outcome(state.scenario)

    def reject(self, consent_id: str) -> None:
        self._require_consent(consent_id).status = "REJECTED"

    def _require_consent(self, consent_id: str) -> _MockConsentState:
        state = self._consents.get(consent_id)
        if state is None:
            raise SahajError("INVALID_INPUT", field="consent_id")
        return state

    def _to_adapter_consent(self, state: _MockConsentState) -> AdapterConsent:
        return AdapterConsent(
            id=state.id,
            status=state.status,
            redirect_url=f"https://sandbox.sahaj.local/consents/{state.id}/approve",
            created_at=state.created_at,
            expires_at=state.expires_at,
        )
